modules/combined/test/tests/contact_verification/hertz_cyl/half_symm_q4/hertz_cyl_half_1deg_template1.i

```diff
@@ -10,7 +10,8 @@
 [Problem]
   type = ReferenceResidualProblem
   solution_variables = 'disp_x disp_y'
-  reference_residual_variables = 'saved_x saved_y'
+  extra_tag_vectors = 'ref'
+  reference_vector = 'ref'
```

Comment:
Almost seems like `extra_tag_vectors` and `reference_vector` can be automated. In general, `extra_tag_vectors=ref` is added to kernels and constrains, and not BCs. Is there cases where this is not true? Could `extra_tag_vectors` automatically use `ref` if `type=ReferenceResidualProblem`?

Certainly, `solution_variables` can be automated? Just thinking about trying to make `ReferenceResidualProblem` easier in future PRs.

Reply:
@bwspenc are there cases where the old `save_in` method breaks from the kernels/constrains and not bcs pattern?

Reply:
@tophmatthews Typically we just put it in the kernels. If you save into the reference vector for the kernels and constraints, the only nonzero values in the vector would be the reactions at the nodes when the solution is converged. 

Do we actually put it in the constraints?  I don't see where that is done in this input, for example. I've been a little out of the loop on the recent changes here. I don't think we have an option in the Contact action to specify a tag vector. I don't think we ever did that with save_in either.  If we don't put it in the constraints, I guess we'd also get the contact forces in the vector, which probably isn't a bad thing. That would probably behave better in cases where you have contact but nothing applied at BCs.

The whole idea of this approach is that it's giving you a very rough measure of something to converge relative to. There's not really a strictly right or a wrong way to do it.

Reply:
@tophmatthews I definitely like the idea of streamlining this. I guess the challenge is that you have to specify the tag vector both in the Action that sets up the kernels and the ReferenceResidualProblem block.

I was hoping that eventually the stuff in ReferenceResidualProblem could just be rolled into FEProblem. It's definitely good to have it moved into the framework, but we still have the problem that if for some reason someone needs to use another Problem type, they can't use this technique, which I think should really be our default way of checking convergence.

Reply:
Why do you have to specify the tag vector? Can't a default name be used, and the kernels check the problem type?

Reply:
Rolled into FEProblem would be great...but my guess is that'd be a battle. I'm just happy it's finally in the framework!

Reply:
> Almost seems like `extra_tag_vectors` and `reference_vector` can be automated. In general, `extra_tag_vectors=ref` is added to kernels and constrains, and not BCs. Is there cases where this is not true? Could `extra_tag_vectors` automatically use `ref` if `type=ReferenceResidualProblem`?
> 
> Certainly, `solution_variables` can be automated? Just thinking about trying to make `ReferenceResidualProblem` easier in future PRs.

The only automation I would support is a default parameter value for `reference_vector` for the `ReferenceResidualProblem`. Here's why I don't support the others:

- `extra_tag_vectors`: this parameter is inherited from `FEProblemBase` and the user could want to create many of these, totally unrelated to convergence critera for `ReferenceResidualProblem`
- `extra_vector_tags`: this parameter is inherited from the `TaggingInterface` and again the user could want to specify many of these, not just for use for reference residuals. Moreover, other physics may prefer to specify saving into the reference residual vector from BCs and ***not*** kernels, just like in my test case. Indeed the [`ReferenceResidualProblem` documentation](https://mooseframework.org/source/problems/ReferenceResidualProblem.html) suggests that integrated fluxes can be a good choice for a reference quantity.

Reply:
`solution_variables`? That's the one that always catches me...

Reply:
Also, almost seems like there could be a `referance_residual_tag` alongside the others...but I get the point

Reply:
> `solution_variables`? That's the one that always catches me...

As the class stands now, this could be automated since it actually errors out if the size of `solution_variables` does not match the number of variables in the non-linear system. ***But*** I want to remove that error since the relative convergence criteria in `ReferenceResidualProblem` doesn't make sense for variables that only have residual contributions from one object, like in the case of the Lagrange Multiplier variable for mechanical or thermal contact.

Reply:
Anyways explicit specification of all these parameters should certainly be allowed. If someone wants to submit PRs providing behind-the-scenes automation if the parameters are not provided, then I would review them 😄

Reply:
BTW @tophmatthews I know you have been using `ReferenceResidualProblem` with mortar contact. I would be curious to see a solve history; the only way I can imagine this working is through an absolute tolerance...?

Reply:
You're assuming I know what I'm doing there...not sure I'm doing it right. My Lagrange multiplier for thermal contact has a terrible norm once in mechanical contact. I use save_in for temperature and displacements, but assign nothing to the Lagrange multiplier constraint. I think that's the right way no?

Reply:
Well your reference quantity will be zero...it's pretty hard to reduce that value by 6-8 orders of magnitude 😃 So I think if you checked why you are achieving convergence, it has to be through an absolute tolerance. If you set your `nl_abs_tol = 0`, I think perhaps you would never see convergence...

Reply:
I assumed if the reference quantity is zero, and your relative tolerance is 1e-6, then if the norm was less than 1e-6 it would be okay?

Reply:
This is the convergence check:
```c++
((_group_resid[i] < _group_ref_resid[i] * rtol) || (_group_resid[i] < abstol))
```
If `group_ref_resid` is zero, then `group_ref_resid * rtol` will also be zero, hence the LHS boolean will always evaulate to `false`.

Reply:
Couldn't we make it so that `solution_variables` is optional? If you don't specify it, it would just default to using the full set of nonlinear variables. I originally had that be specified so that the correspondence of reference variables to solution variables was defined, but that's probably not needed anymore with the tagging option.

Reply:
In the parameter sense it is currently optional, e.g. it is not a required parameter. However, you'd error out if you didn't specify it currently. I guess I can add that to this PR 🙄

Reply:
It probably makes sense to handle that separately.  These tests already specify `solution_variables`. Maybe in a separate PR we can make that optional and then go remove `solution_variables` from all the tests.

Reply:
Maybe I don't fully understand still...but isn't `_group_ref_resid[i]` just moving the "zero" to determine convergence? In my head, it seems that it should look more like
```
_group_resid[i] - _group_ref_resid[i] < rtol
```
I guess that is more like absolute tolerance then...

Reply:
For what you just wrote, you would have "convergence" when `_group_resid == _group_ref_resid`, e.g. when we haven't converged at all.

Relative tolerance is always based on `current_residual/initial_residual < rtol` or `current_residual/reference_residual < rtol`

Reply:
Yeah, you're right. Perhaps the reference should be 1 if nothing is added? This becomes similar to absolute tolerance check, but still allows each variable to be evaluated separately.

Reply:
It seems to me, in the absence of any `extra_vector_tags` for a given variable, setting `_group_ref_resid=1` will allow the 1) use of `type=ReferenceResidualProblem` and convergence to the relative tolerance, 2) comparison of the variable to its own norm (as opposed to the absolute tolerance check).

Otherwise, for thermal contact, the Lagrange multiplier will always prevent rel_tol convergence if using `ReferenceResidualProblem`, right?

Reply:
you could have a scenario where the overall L2 norm is greater than the absolute tolerance, but if all your solution variables other than the LM variable  are converged relatively and the LM variable is converged absolutely, then you would have a “relative converged” outcome

Reply:
Yes...that is true. Kinda that the absolute tolerance comes in there, but totally true.